bzip2 1.0.8: Rebuild on win

skupr-anaconda opened the pull request:
**Destination channel:** main

### Links

- [PKG-4142](https://anaconda.atlassian.net/browse/PKG-4142) 
- [Upstream repository](https://sourceware.org/git/?p=bzip2.git;a=tree;hb=9de658d248f9fd304afa3321dd7a9de1280356ec)

### Explanation of changes:

- Set empty `BZIP2/BZIP` variables in `bld.bat` to avoid an error because the default value is `1.0` in [cbc.yaml](https://github.com/AnacondaRecipes/aggregate/blob/0de119406a04fb8ba753524c6ac035c75c1fe2a7/conda_build_config.yaml#L16)
- Replace `jom` by `nmake` as we do not need multiprocessing for this small package
- Clean up and rewrite the `bld.bat`
- Copy executables (e.i., `bzip2.exe`) to `%LIBRARY_BIN%`
- Add a patch to fix the `dll` creation in `makefile.msc` (see the [original one](https://sourceware.org/git/?p=bzip2.git;a=blob;f=makefile.msc;h=799a18a5f1a7b084274ce1cd63997aeb4f9c2688;hb=9de658d248f9fd304afa3321dd7a9de1280356ec)) and remove outdated `DESCRIPTION` in `libbz2.def`
- Clean up the `meta.yaml`
- Bump the build number to `1`
- Remove redundant `m2-gcc-libs` 
- Add [recipe/bz2.c](https://github.com/AnacondaRecipes/bzip2-feedstock/pull/3/files#diff-2d3da61d2f87859b584ee8c5b99325b784c27a3695e4bc9e7a321d2396d10bf3) for the tests
- Add `{{ compiler('c') }}` to `test/requires` to compile `bz2.c` in tests
- Move the test commands for `win` to `run_test.bat`
- Update `home`
- Update `doc_url`
- Add `dev_url`
- Update the `license` name as `bzip2-1.0.8`. The [spdx](https://spdx.org/licenses/) doesn't have this name because only `bzip2-1.0.6` [exists](https://spdx.org/licenses/bzip2-1.0.6.html) but it's incorrect. The real name must be `bzip2-1.0.8`, see the upstream [LICENSE](https://sourceware.org/git/?p=bzip2.git;a=blob;f=LICENSE;h=81a37eab7a5be1a34456f38adb74928cc9073e9b;hb=9de658d248f9fd304afa3321dd7a9de1280356ec#l40) file


### Notes:


- We noticed an anomaly when `R` packages tried to decompress `R` **data** files during the installation of `R` packages if we used `bzip2` on `win-64` from the defaults channel. The installation of R packages dies without any valuable error messages like:
- 
  ```
          ** building package indices
         ERROR: installing package indices failed
   ```
- 
  ```
          ** R
          ** data
          ** inst
          ** byte-compile and prepare package for lazy loading"```
- We debugged the `R` source code and found that the root cause was the incorrectly built `bzip2` on `win64`.

- Should we **test** other downstream packages (`python`, `libarchive`, `libzip`, `libmamba`, see the [list](https://metayaml-conda.fly.dev/metayaml/reverse_dependencies?name=bzip2)) from defaults to verify that our changes do not break them? 
  - We only tested it with `r-base 4.3.1` and 800 of `R` packages and we can say that the issue was resolved.

[PKG-4142]: https://anaconda.atlassian.net/browse/PKG-4142?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

skupr-anaconda committed:
Rebuild on win

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_a3psu94uwm/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:77: http_url: http://www.bzip.org/ is not https
- clone/recipe/meta.yaml:70: incorrect_license: The recipe's `about/license` key is not an SPDX compliant license or license exception, reference https://spdx.org/licenses/
- clone/recipe/meta.yaml:77: http_url: http://www.bzip.org/docs.html is not https
- clone/recipe/meta.yaml:77: missing_dev_url: The recipe is missing a dev_url
===== Final Report: =====
4 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Rebuild on win

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_56vv34_763/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:77: http_url: http://www.bzip.org/ is not https
- clone/recipe/meta.yaml:70: incorrect_license: The recipe's `about/license` key is not an SPDX compliant license or license exception, reference https://spdx.org/licenses/
- clone/recipe/meta.yaml:77: http_url: http://www.bzip.org/docs.html is not https
- clone/recipe/meta.yaml:77: missing_dev_url: The recipe is missing a dev_url
===== Final Report: =====
4 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Rebuild on win

skupr-anaconda committed:
Fix license

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_5dqcfdhjyw/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:77: missing_dev_url: The recipe is missing a dev_url
- clone/recipe/meta.yaml:77: http_url: http://www.bzip.org/ is not https
- clone/recipe/meta.yaml:77: http_url: http://www.bzip.org/docs.html is not https
===== Final Report: =====
3 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Add a test file and test it

skupr-anaconda committed:
Fix the test

skupr-anaconda committed:
Move windows tests to run_test.bat

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_cbqvi9yu9_/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:61: missing_tests: No tests were found.
- clone/recipe/meta.yaml:61: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Rewrite the test command

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_ff0ww1bjkr/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:61: missing_tests: No tests were found.
- clone/recipe/meta.yaml:61: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Rewrite the test command

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_e4rc4srekd/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:61: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:61: missing_tests: No tests were found.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Rewrite the test command

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_fc47a03r4p/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:61: missing_tests: No tests were found.
- clone/recipe/meta.yaml:61: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Add run_test.py

skupr-anaconda committed:
Add run_test.py

skupr-anaconda committed:
Fix the test command

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_ffb4faa8h8/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Fix the test command

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_8anbqjez4w/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Debug bld.bat

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_439unsk9pt/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Cope exes

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_e27wizgccs/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Test bzip2.exe

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_c025j7t746/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Test bzip2 -dc

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_67f08z6ml0/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_afd0rn0giu/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Fix tests

skupr-anaconda committed:
Add patches

skupr-anaconda committed:
Add trailing whitespaces to patches, copy libbz2.exp

skupr-anaconda committed:
Add one patch

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_404m5t84ig/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Set BZIP2/BZIP

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_47j0ev0swg/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Copy LIBBZ2.dll to LIBRARY_BIN

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_61yz3pbzo9/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Make out:LIBBZ2.dll

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_b5n67habz4/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Add abs.yaml

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_5b70e755fl/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Fix abs.yaml

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_33h9dec1q4/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Fixes

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_b7q98j0uy_/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Remove an obj file

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_3dwtlji74_/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Fix dll

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_b0u9x4uhrt/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Fix dll

skupr-anaconda committed:
Remove dll

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_9caxsdgorm/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Use /dll /subsystem:WINDOWS

skupr-anaconda committed:
Fix dll

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_ecqqv5vbci/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Fix header

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_f6oc1p7_t0/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Fix header

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_d8ne81w0vs/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Add options

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_bb3mrpkurr/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Rename lib

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_c9i31f12kg/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Rewrite commands

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_f577y8jx7l/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Rewrite commands

skupr-anaconda committed:
Rewrite command

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_c9277bhqgb/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Rewrite command

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_70yikn745v/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Rewrite command

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_bbn_clmd2s/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Fix a patch and bld.bat

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_824c73r7u9/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Debug

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_0b4fok1dbl/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Debug

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_5eg9pob_65/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Debug

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_3ckajdq0ge/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Debug

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_dai5e5ijvl/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Add /MD to link with MSVCRT.lib

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_29o86zt11y/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Add /MD to link with MSVCRT.lib

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_690_8x9mlx/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

skupr-anaconda committed:
Clean up the debugging code

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_61lkbn0y1i/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:67: incorrect_license: The recipe's `about/license` key is not an SPDX compliant license or license exception, closest match: bzip2-1.0.6
- clone/recipe/meta.yaml:61: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:61: missing_tests: No tests were found.
===== Final Report: =====
3 Errors and 0 Warnings were found

</details>

ifitchet commented on recipe/bld.bat:
```diff
@@ -1,20 +1,32 @@
-REM edited from the one from Anaconda
+@echo on
+
+REM Set empty variables to avoid an error because by defaults the value is "1.0",
+REM see https://github.com/AnacondaRecipes/aggregate/blob/0de119406a04fb8ba753524c6ac035c75c1fe2a7/conda_build_config.yaml#L16
```
We could clarify that:

```suggestion
REM The bzip2 executable uses the environment variables BZIP2 and BZIP to initialise its ArgList
REM conda sets these values from conda_build_config.yaml and bzip2 tries to open a file called "1.0"
```

ifitchet commented on recipe/bld.bat:
```diff
@@ -1,20 +1,32 @@
+copy LIBBZ2.dll %LIBRARY_BIN%\LIBBZ2.dll || exit 1
+
+REM Copy exe files
+copy *.exe %LIBRARY_BIN% || exit 1
```
For windows we never used to copy the executable(s).

In general (Windows and elsewhere) the corresponding commands, `bunzip2`, `bzcat` etc., are simply renamed versions of the same executable.

Should we be creating those for consistency?

skupr-anaconda replied:
It's a good question. I'm not sure if there are use-cases on Windows for that

skupr-anaconda committed:
Update recipe/bld.bat

Co-authored-by: Ian Fitchet <123584373+ifitchet@users.noreply.github.com>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_f8xune8kn0/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:67: incorrect_license: The recipe's `about/license` key is not an SPDX compliant license or license exception, closest match: bzip2-1.0.6
- clone/recipe/meta.yaml:61: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:61: missing_tests: No tests were found.
===== Final Report: =====
3 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Add --no-error-overlinking to abs.yaml

skupr-anaconda committed:
Add the reason to not use --error-overlinking

JeanChristopheMorinPerso commented on recipe/build_dll.patch:
```diff
@@ -0,0 +1,87 @@
++++ b/libbz2.def
+@@ -1,5 +1,4 @@
+ LIBRARY			LIBBZ2
+-DESCRIPTION		"libbzip2: library for data compression"
```
Why is this removed?

skupr-anaconda replied:
It's an outdated field, the compiler don't support it anymore

JeanChristopheMorinPerso reviewed (commented):
Considering the amount of problems this package has caused you, I feel there should be much more comments to describe what makes it work now compared to before. As things are today, I don't feel confidant at all that we can sustainably maintain this package when future versions will be released without fearing to break the msys2 work.

skupr-anaconda commented:
I've added to `abs.yaml`:
```
build_parameters:
  # See https://github.com/conda/conda-build/issues/3650
  - "--no-error-overlinking"
```

skupr-anaconda commented:
> Considering the amount of problems this package has caused you, I feel there should be much more comments to describe what makes it work now compared to before. As things are today, I don't feel confidant at all that we can sustainably maintain this package when future versions will be released without fearing to break the msys2 work.

I found these [patches](https://github.com/philr/bzip2-windows/tree/v1.0.8.0/patches) from the `bzip2-windows` project and by trial and error @ifitchet and I finally can compile C code with the `bzip2` library (using `gcc` and `cl.exe`). 
Our goal is to use our `bzip2` package on `win64` to build R packages and **not** `ucrt64-bzip2` (from [MSYS2](https://packages.msys2.org/package/mingw-w64-ucrt-x86_64-bzip2?repo=ucrt64)).

My understanding is that the next release of `bzip2` will be `1.1+`, see the new successor https://gitlab.com/bzip2/bzip2/-/tree/master?ref_type=heads#maintainers. It will have better Windows support because it relies on CMake.

@ifitchet Should we use `bzip2 1.0.8` in `r-base`'s **host** to prevent any compatibility issues?

JeanChristopheMorinPerso commented:
Ok thank you for the additional details @skupr-anaconda. Since you took patches (and modified some) from another repo, we need to give attribution to the author and potentially link to where they come from. Note that the patches are also under a different license (I think) than bzip2. We also have to show this.

I would also add somewhere in the recipe or the patches that they might not be requied in future versions.

skupr-anaconda commented:
> Ok thank you for the additional details @skupr-anaconda. Since you took patches (and modified some) from another repo, we need to give attribution to the author and potentially link to where they come from. Note that the patches are also under a different license (I think) than bzip2. We also have to show this.
> 
> I would also add somewhere in the recipe or the patches that they might not be required in future versions.

The [recipe/build_dll.patch](https://github.com/AnacondaRecipes/bzip2-feedstock/pull/3/files#diff-8453442a592179c96354f7bf02d91270697f78e93aa0515e9360c43f7646c39d) I made personally, it only partially based on  the bzip2-windows's vision how to fix the issue (because the bzi2-windows's patch doesn't work). Mainly, Ian and I  after debugging have found the correct way to patch the upstream `makefile.msc`.

I'll add a comment that this patch might not be required in future versions

skupr-anaconda committed:
Add a comment about build_dll.patch

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_f6zct_8dq6/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
- clone/recipe/meta.yaml:63: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:69: incorrect_license: The recipe's `about/license` key is not an SPDX compliant license or license exception, closest match: bzip2-1.0.5
===== Final Report: =====
3 Errors and 0 Warnings were found

</details>

JeanChristopheMorinPerso commented:
@skupr-anaconda The license of bzip2-windows repo clearly states:
>The above copyright notice and this permission notice shall be included in all 
copies or substantial portions of the Software.

https://github.com/philr/bzip2-windows/blob/v1.0.8.0/patches/01-replace_deprecated_compile_options.diff was taking as is, https://github.com/philr/bzip2-windows/blob/v1.0.8.0/patches/02-build_dll.diff almost as is, and https://github.com/philr/bzip2-windows/blob/v1.0.8.0/patches/04-remove_unsupported_description_option.diff as is too. We must keeop copyright and license on such work.

skupr-anaconda commented:
> @skupr-anaconda The license of bzip2-windows repo clearly states:
> 
> > The above copyright notice and this permission notice shall be included in all
> > copies or substantial portions of the Software.
> 
> https://github.com/philr/bzip2-windows/blob/v1.0.8.0/patches/01-replace_deprecated_compile_options.diff was taking as is, https://github.com/philr/bzip2-windows/blob/v1.0.8.0/patches/02-build_dll.diff almost as is, and https://github.com/philr/bzip2-windows/blob/v1.0.8.0/patches/04-remove_unsupported_description_option.diff as is too. We must keeop copyright and license on such work.

So what is the correct way to license it in our case? Should we add the license name `MIT` to `license` and add `Copyright (c) 2015-2019 Philip Ross` to the description of the patch? Because we do not copy files.

JeanChristopheMorinPerso commented:
>Because we do not copy files

But you copied the content of them as is. Same thing than copying. That's basically the point I'm trying to make. At the very least we need to say where these bits of patches come from or from what they were inspired/modified.

cbouss commented on recipe/meta.yaml:
```diff
@@ -1,32 +1,38 @@
 
 build:
-  number: 0
+  number: 1
```
```suggestion
  number: 5
```

The highest build number for 1.0.8 is b4 for osx-arm64.

cbouss replied:
Can you check the differences between the recipes in the latest bzip packages available for 1.0.8 and what is in this feedstock?

skupr-anaconda replied:
linux-64 <-> linux-aarch64:
```
╰─❯ diff linux-64_bzip2-1.0.8-h7b6447c_0/info/recipe/meta.yaml.template linux-aarch64_bzip2-1.0.8-hfd63f10_2/info/recipe/meta.yaml.template
14c14
<   number: 0
---
>   number: 2
20a21,23
>     - m2-patch  # [win]
>     - m2-gcc-libs  # [win]
>     - patch  # [not win]
```
linux-64 <-> osx-arm64:
about the creation of the osx.arm64's meta.yaml:
```
# This file created by conda-build 3.21.4
# meta.yaml template originally from:
# /Users/ktietz/Code/oss/cf-packages/bzip2-feedstock/recipe, last modified Sun Jun 20 20:52:29 2021
```
```
╰─❯ diff linux-64_bzip2-1.0.8-h7b6447c_0/info/recipe/meta.yaml.template osx-arm64_bzip2-1.0.8-h620ffc9_4/info/recipe/meta.yaml.template
14c14
<   number: 0
---
>   number: 4
59c59
<   license: bzip2
---
>   license: bzip2-1.0.6
71d70
<     - jjhelmus
```

```
diff linux-64_bzip2-1.0.8-h7b6447c_0/info/recipe/bld.bat osx-arm64_bzip2-1.0.8-h620ffc9_4/info/recipe/bld.bat
3a4
> nmake -f makefile.msc
```

```
╰─❯ diff linux-64_bzip2-1.0.8-h7b6447c_0/info/recipe/bld.bat osx-arm64_bzip2-1.0.8-h620ffc9_4/info/recipe/build.sh
1,20c1,19
< REM edited from the one from Anaconda
<
< REM Build step
< jom -f makefile.msc lib bzip2
< if errorlevel 1 exit 1
<
< REM Install step
< copy libbz2.lib %LIBRARY_LIB%\libbz2_static.lib || exit 1
< REM Some packages expect 'bzip2.lib', so make another copy
< copy libbz2.lib %LIBRARY_LIB%\bzip2_static.lib || exit 1
< copy bzlib.h %LIBRARY_INC% || exit 1
<
< cl /O2 /Ibzlib.h /Ibzlib_private.h /D_USRDLL /D_WINDLL blocksort.c bzlib.c compress.c crctable.c decompress.c huffman.c randtable.c /LD /Felibbz2.dll /link /DEF:libbz2.def
< if errorlevel 1 exit 1
<
< copy libbz2.lib %LIBRARY_LIB% || exit 1
< REM Some packages expect 'bzip2.lib', so make copies
< copy libbz2.lib %LIBRARY_LIB%\bzip2.lib || exit 1
< copy libbz2.dll %LIBRARY_BIN% || exit 1
< copy libbz2.dll %LIBRARY_BIN%\bzip2.dll || exit 1
---
> #!/bin/sh
>
> set -ex
>
> # toolchain flags + bzip flags + fpic
> export CFLAGS="${CFLAGS} -Wall -Winline -O2 -g -D_FILE_OFFSET_BITS=64 -fPIC"
> USED_CC=${GCC:-${CC}}
> make install PREFIX=${PREFIX} CFLAGS="${CFLAGS}" CC="${USED_CC}"
>
> # build shared library
> if [[ ${target_platform} =~ .*linux.* ]]; then
>   make -f Makefile-libbz2_so CFLAGS="${CFLAGS}" CC="${USED_CC}"
>   ln -s libbz2.so.${PKG_VERSION} libbz2.so
>   cp -d libbz2.so* ${PREFIX}/lib/
> elif [[ ${target_platform} == osx-* ]]; then
>   ${USED_CC} -shared -Wl,-install_name -Wl,libbz2.dylib -o libbz2.${PKG_VERSION}.dylib blocksort.o huffman.o crctable.o randtable.o compress.o decompress.o bzlib.o ${LDFLAGS}
>   cp libbz2.${PKG_VERSION}.dylib ${PREFIX}/lib/
>   ln -s libbz2.${PKG_VERSION}.dylib ${PREFIX}/lib/libbz2.dylib
> fi
```

I'll reflect on those changes by:
- adding `export CFLAGS="${CFLAGS} -Wall -Winline -O2 -g -D_FILE_OFFSET_BITS=64 -fPIC"` to `build.sh`
- adding `${LDFLAGS}` for `osx-*`

skupr-anaconda replied:
I modified the build.sh

skupr-anaconda committed:
Modify:
- Add a notice about the origin of the patch and its license.

- Bump the build number to 5 because osx-arm64 has the number 4.

- Add CFLAGS to build.sh and LDFLAGS for osx-*.

- Add a comment about the license name

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_b4q2ka24k7/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:72: incorrect_license: The recipe's `about/license` key is not an SPDX compliant license or license exception, closest match: bzip2-1.0.6
- clone/recipe/meta.yaml:63: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:63: missing_tests: No tests were found.
===== Final Report: =====
3 Errors and 0 Warnings were found

</details>

skupr-anaconda commented:
> > Because we do not copy files
> 
> But you copied the content of them as is. Same thing than copying. That's basically the point I'm trying to make. At the very least we need to say where these bits of patches come from or from what they were inspired/modified.

I've added info about the origin and the license of the patch.

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -43,30 +49,27 @@ test:
-    - if not exist %LIBRARY_LIB%\\libbz2_static.lib exit 1  # [win]
+    - test -f ${PREFIX}/include/bzlib.h       # [unix]
+    - test -f ${PREFIX}/lib/libbz2.a          # [unix]
+    - test -f ${PREFIX}/lib/libbz2.so         # [linux]
+    - test -f ${PREFIX}/lib/libbz2.dylib      # [osx]
```
```suggestion
    - test -f ${PREFIX}/lib/libbz2.${SHLIB_EXT}        # [unix]
```

skupr-anaconda replied:
It will lead to an error because of `/lib/libbz2..so`. I'll fix it

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -1,32 +1,38 @@
+  requires:                # [win]
+    - {{ compiler('c') }}  # [win]
+  commands:                                 # [unix]
+    - bzip2 --help                          # [unix]
```
```suggestion
    - bzip2 --help
```

skupr-anaconda replied:
Funny but if your `meta.yaml` has this wrongly placed selector `# [unix]`:
```
test:
  commands:                                 # [unix]
    - bzip2 --help
```
Then  you will get an error in the host environment on `win-64`:
`InvalidVersionSpec: Invalid version '--help': empty version component`

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -43,30 +49,27 @@ test:
     {% for executable in bzip2_executables %}
     - test -f ${PREFIX}/bin/{{ executable }}  # [unix]
```
```suggestion
    - test -f ${PREFIX}/bin/{{ executable }}
```

skupr-anaconda replied:
We test on win64 only two executables in `run_test.bat`:
```
if not exist %LIBRARY_BIN%\bzip2.exe exit 1
if not exist %LIBRARY_BIN%\bzip2recover.exe exit 1
```

No other exe files are created. Ian has already asked about this issue https://github.com/AnacondaRecipes/bzip2-feedstock/pull/3#discussion_r1500405484. I'm still not sure if we need them

skupr-anaconda committed:
Update recipe/meta.yaml

Co-authored-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>

skupr-anaconda committed:
Update recipe/meta.yaml

Co-authored-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>

skupr-anaconda committed:
Update recipe/meta.yaml

Co-authored-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>

skupr-anaconda committed:
Remove dot

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_b2on8ognof/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
- clone/recipe/meta.yaml:71: incorrect_license: The recipe's `about/license` key is not an SPDX compliant license or license exception, closest match: bzip2-1.0.5
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
3 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Test executables correctly

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_b8uy7068wo/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: missing_tests: No tests were found.
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:71: incorrect_license: The recipe's `about/license` key is not an SPDX compliant license or license exception, closest match: bzip2-1.0.5
===== Final Report: =====
3 Errors and 0 Warnings were found

</details>

skupr-anaconda committed:
Remove the selector

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_e39ex19mvc/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:71: incorrect_license: The recipe's `about/license` key is not an SPDX compliant license or license exception, closest match: bzip2-1.0.6
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

cbouss reviewed (approved):
Nice work!

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_4drvgxt7mg/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:62: has_run_test_and_commands: Test commands are not executed when run_test.sh (.bat...) is present.
- clone/recipe/meta.yaml:71: incorrect_license: The recipe's `about/license` key is not an SPDX compliant license or license exception, closest match: bzip2-1.0.6
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>
